Custom resolver function to provide routing driver initial server addresses

zhenlineo opened the pull request.

Commit:
Customized resolver function to provide routing driver initial server addresses.

zhenlineo committed:
Fixed the test error due to the mis-use of a mutable static `RoutingTable.Empty` const.
Improved test logging in RoutingDriverIT and Stubserver tests for better debug logging.

Commit:
Fixing .net core 1.1 test failure

lutovich commented on Neo4j.Driver/Neo4j.Driver/Internal/RoutingSettings.cs:
```diff
@@ -17,24 +17,55 @@
+
+    internal interface IInitialServerAddressProvider
+    {
+        ISet<Uri> Resolve();
```
It's a bit strange to have a "provider" with "resolve" methods. Wdyt about renaming methods to `Get` and `GetAsync`?

lutovich reviewed (commented):
Looks good. Made a small comment. It would also be nice to add a stub server test which tests the custom resolver.

ali-ince commented on Neo4j.Driver/Neo4j.Driver/V1/IAddressResolver.cs:
```diff
@@ -0,0 +1,60 @@
+        /// </summary>
+        /// <param name="uri">The server address to resolve.</param>
+        /// <returns>A task of resolved server addresses in a set.</returns>
+        Task<ISet<Uri>> ResolveAsync(Uri uri);
```
Do we really need an async version? I believe mostly people will use it to return a static list of hosts, won't they?

zhenlineo replied:
Do you think this async method would be useful if a user uses some DNS resolver or server address registry service to tell what servers are available? If you could only think of static use cases, then we definitely do not need a async version.

ali-ince replied:
Yeah, maybe - but I believe it will be static most of the time. We can remove the async one here and consider adding it later if anyone asks for it?

ali-ince commented on Neo4j.Driver/Neo4j.Driver/V1/IAddressResolver.cs:
```diff
@@ -0,0 +1,60 @@
+        /// </summary>
+        /// <param name="uri">The server address to resolve.</param>
+        /// <returns>Resolved server addresses in a set.</returns>
+        ISet<Uri> Resolve(Uri uri);
```
How do you feel about not using `Uri` for this? We only want users to map `hostname:port` pairs to another set of `hostname:port` pairs and `Uri` contains more information than that. Wdyt?

ali-ince reviewed (commented):
Looks good. Made a couple of comments about the new public `IServerAddressResolver` interface.

Commit:
Fixing after review.
Added ServerAddress class to only expose Host and Port to user when writing their customer resolver funtion

ali-ince commented on Neo4j.Driver/Neo4j.Driver/V1/IAddressResolver.cs:
```diff
@@ -29,32 +29,113 @@ namespace Neo4j.Driver.V1
+    /// A server address consists of <see cref="ServerAddress.Host"/> and <see cref="ServerAddress.Port"/>.
+    /// This address specifies where the driver to find the server.
+    /// </summary>
+    public sealed class ServerAddress
```
It may worth adding `IEquatable<ServerAddress>` as an implemented interface.

Commit:
IEquatable for the interfaces that provides a strongly typed Equals method
Sealed TxConfig too
